File: knotlink-sdk-qt/knotlink/opensocketresponser.cpp
/*
 * KnotLink SDK - C++
 * Copyright (c) 2024-2026 KnotLink Contributors
 * SPDX-License-Identifier: MIT
 */

#include "opensocketresponser.h"
#include <QDebug>

Q_LOGGING_CATEGORY(knotlinkResponser, "knotlink.responser")

OpenSocketResponser::OpenSocketResponser(QString APPID, QString OpenSocketID, QObject *parent)
    : QObject(parent), appID(APPID), openSocketID(OpenSocketID)
{
    init();
}


void OpenSocketResponser::init()
{
    KLresponser = new TcpClient(this);
    KLresponser->connectToServer("127.0.0.1",6378);
    connect(KLresponser, &TcpClient::receivedData, this, &OpenSocketResponser::dataRecv);
    QString s_key = appID + "-" + openSocketID;
    // 将 s_key 转换为 QByteArray
    QByteArray s_key_bytes = s_key.toUtf8(); // 转换为 UTF-8 编码的 QByteArray
    KLresponser->sendData(s_key_bytes);
}

void OpenSocketResponser::dataRecv(const QByteArray &data)
{
    QString s_data = QString::fromUtf8(data);

    QString delimiter = "&*&"; // 分隔符
    QStringList parts = s_data.split(delimiter); // 按分隔符分割字符串

    if (parts.size() < 2) {
        qCWarning(knotlinkResponser) << "Invalid data format, missing delimiter:" << delimiter;
        return;
    }

    QString key = parts[0];
    // payload 是分隔符之后的所有内容（payload 本身可能包含 &*&）
    QString t_data = s_data.mid(key.length() + delimiter.length());

    emit receivedData_a(t_data.toUtf8(),key);
    emit receivedData(t_data,key);
}

void OpenSocketResponser::sendBack(const QString &data,QString questionID)
{
    sendBack(data.toUtf8(),questionID);
}

void OpenSocketResponser::sendBack(const QByteArray &data, QString questionID)
{
    QByteArray result = questionID.toUtf8() + "&*&" + data;
    KLresponser->sendData(result);
}

File: knotlink-sdk-pyqt/knotlink-pyqt/knotlink-pyqt/OpenSocketResponser.py
# KnotLink SDK - Python
# Copyright (c) 2024-2026 KnotLink Contributors
# SPDX-License-Identifier: MIT

from PyQt5.QtCore import QObject, pyqtSignal
from .tcpclient import TcpClient  

class OpenSocketResponser(QObject):
    receivedData = pyqtSignal(str, str)
    receivedData_a = pyqtSignal(bytes, str)

    def __init__(self, APPID: str, OpenSocketID: str, parent=None):
        super().__init__(parent)
        self.appID = APPID
        self.openSocketID = OpenSocketID
        self.init()

    def init(self):
        self.KLresponser = TcpClient(self)
        self.KLresponser.connectToServer("127.0.0.1", 6378)
        self.KLresponser.receivedData.connect(self.dataRecv)
        s_key = f"{self.appID}-{self.openSocketID}"
        s_key_bytes = s_key.encode("utf-8")  # 转换为 UTF-8 编码的 QByteArray
        self.KLresponser.sendData(s_key_bytes)

    def dataRecv(self, data: bytes):
        s_data = data.decode("utf-8")

        delimiter = "&*&"  # 分隔符
        parts = s_data.split(delimiter)  # 按分隔符分割字符串

        if len(parts) != 2:  # 确保分割后有两部分
            print(f"Invalid data format. Expected two parts separated by {delimiter}")
            return

        key = parts[0]  # 前一部分作为 key
        t_data = parts[1]  # 后一部分作为 t_data
        print(f"Key: {key}")
        print(f"t_data: {t_data}")

        self.receivedData_a.emit(t_data.encode("utf-8"), key)
        self.receivedData.emit(t_data, key)

    def sendBack(self, data: str, questionID: str):
        self.sendBack_a(data.encode("utf-8"), questionID)

    def sendBack_a(self, data: bytes, questionID: str):
        data_r = f"{questionID}&*&{data.decode('utf-8')}"
        self.KLresponser.sendData(data_r.encode("utf-8"))
